bindings/pyroot/cppyy/cppyy-backend/CMakeLists.txt

```diff
@@ -45,6 +45,12 @@ foreach(val RANGE ${how_many_pythons})
                              RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT libraries
                              LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT libraries
                              ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT libraries)
+  if (NOT CMAKE_INSTALL_LIBDIR STREQUAL CMAKE_INSTALL_PYTHONDIR)
+    # add a symlink to ${libname} in CMAKE_INSTALL_PYTHONDIR
+    install(CODE "execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink\
+            ${CMAKE_INSTALL_FULL_LIBDIR}/lib${libname}.so\
+            ${CMAKE_INSTALL_FULL_PYTHONDIR}/lib${libname}.so)")
+  endif()
```

Comment:
This assumes `-Dsoversion=OFF` and will miss links to the actual library (i.e. `.so.6.xx.yy` extension). It will likely not work with `-Dsoversion=ON` which should be used when distributing ROOT binaries to avoid mixing up versions at runtime.

Reply:
Also, as I've mentioned before, it would be much better to fix the backend library loader from cppyy to look into ROOT's libdir as well instead of creating symlinks. This only hides the root cause, which is that the loader fails to find the library. There is [code for it already in cppyy](https://github.com/root-project/root/blob/7987937edab455b6e9119e887e09de522bbfab63/bindings/pyroot/cppyy/cppyy-backend/cling/python/cppyy_backend/loader.py#L21-L47), it just needs some adjustments to find the libraries.

Reply:
@etejedor I cannot make a decision on this, up to you

Reply:
I would try to avoid patching cppyy, if possible. That brings us further away from building on top of upstream cppyy. Is there any strong reason against using the symlinks?

Reply:
Well, the versioned links are a good reason. Also, cppyy is already patched in any case, and this is a patch that should be easy to upstream, as it's needed by cppyy itself. All that's needed is to check in `root-config --libdir` when loading libraries, should be an easy patch.

Reply:
IIUC the patch you are proposing would make cppyy look into the ROOT lib dir for its libraries. I don't see why such a patch would be needed by cppyy and it is very unlikely that it is accepted upstream.

Could you elaborate more on the versioned links reason?

Reply:
When you use `-Dsoversion=ON` the real library file is e.g., `libXXX.so.6.23.01` so just linking the `.so` may cause failures.

Reply:
> I don't see why such a patch would be needed by cppyy

Cppyy needs to load its "mini ROOT", including libCore (see my link from above). So it does need to look into ROOT library directory, especially if that's where `libcppyy_backend` is going to be installed. Enrico also mentioned that cppyy cannot find headers if installed in `/usr/include/ROOT`, which is another thing that may be worth investigating and fixing.

Anyway, since my comments don't seem to have any effect in changing your minds from symlinking, I will just get out of the way and let you guys decide what you prefer to do.

Reply:
(I tried to use `$<TARGET_FILE_NAME:${libname}>` instead of hard-coding `lib${libname}.so` but cmake is very picky about where generator expressions can appear and I could not make it work)

Reply:
Patching cppyy vs symlinking: I don't have a strong opinion.

Making sure things work with `-Dsoversion=ON`: I think it's a must before this gets merged.
So I checked. With `-Dsoversion=ON`, the situation on Debian is the following:

```
root@b3180b9a8bd1:~# ls /usr/lib/x86_64-linux-gnu/libcppyy*
/usr/lib/x86_64-linux-gnu/libcppyy3_7.so       /usr/lib/x86_64-linux-gnu/libcppyy3_7.so.6.23.01  /usr/lib/x86_64-linux-gnu/libcppyy_backend3_7.so.6.23
/usr/lib/x86_64-linux-gnu/libcppyy3_7.so.6.23  /usr/lib/x86_64-linux-gnu/libcppyy_backend3_7.so  /usr/lib/x86_64-linux-gnu/libcppyy_backend3_7.so.6.23.01
```

so `libcppyy3_7.so` is _also_ there.

Reply:
> Cppyy needs to load its "mini ROOT", including libCore (see my link from above). So it does need to look into ROOT library directory, especially if that's where libcppyy_backend is going to be installed.

When installed alone, Cppyy puts its mini ROOT libraries (libCore.so, etc.) in `site-packages/cppyy_backend/lib`, and that's where cppyy looks. Our case is different, since those libraries are ours and they are installed elsewhere (not in the Python directories). But that case is not relevant for cppyy, that is why I said the patch would not make it upstream.

> Enrico also mentioned that cppyy cannot find headers if installed in /usr/include/ROOT, which is another thing that may be worth investigating and fixing.

There is no reason why vanilla cppyy would need to find its headers in `/usr/include/ROOT`, since that is not a standard directory. The headers it needs (and that it installs) are not ROOT's, and it looks for them in `/usr/include/CPyCppyy`. A solution would be to install them there.

It is not at all that we don't want to change our mind about the symlinks, but this is a solution that can potentially work and we are discussing its pros/cons. I appreciate your feedback as always.

Reply:
`-Dsoversion=ON` is not broken as far as I can tell (tested on debian stable), and from what I gather from the conversation patching cppyy to find things in ROOT directories is not desired. marking this as resolved.